contracts/governance/liquidation/LiquidationStrategy.sol

```diff
@@ -0,0 +1,171 @@
+  }
+
+  LiquidationSchedule public liquidationSchedule;
+  ITreasuryPool public feePool;
+  address payable public treasuryPool;
+  // list of tokens that can be liquidate to
+  mapping(address => bool) public whitelistedTokens;
+
+  event FeePoolSet(address indexed feePool);
+  event TreasuryPoolSet(address indexed treasuryPool);
+  event LiquidationScheduleUpdated(uint128 startTime, uint64 repeatedPeriod, uint64 duration);
+  event WhitelistedTokenUpdated(address indexed token, bool indexed isAdd);
+
+  modifier onlyWhenLiquidationEnabled() {
+    require(isLiquidationEnabled(), 'only when liquidation enabled');
+    _;
+  }
+
+  constructor(
+    address _admin,
+    address _feePool,
+    address payable _treasuryPool,
+    uint128 _startTime,
+    uint64 _repeatedPeriod,
+    uint64 _duration,
+    address[] memory _whitelistedTokens
+  ) PermissionAdmin(_admin) {
+    _setFeePool(_feePool);
+    _setTreasuryPool(_treasuryPool);
+    _setLiquidationSchedule(_startTime, _repeatedPeriod, _duration);
+    _updateWhitelistedToken(_whitelistedTokens, true);
+  }
+
+  function updateLiquidationSchedule(
+    uint128 _startTime,
+    uint64 _repeatedPeriod,
+    uint64 _duration
+  )
+    external onlyAdmin
+  {
+    _setLiquidationSchedule(_startTime, _repeatedPeriod, _duration);
+  }
+
+  function updateFeePool(address _feePool) external onlyAdmin {
+    _setFeePool(_feePool);
+  }
+
+  function updateTreasuryPool(address payable _treasuryPool) external onlyAdmin {
+    _setTreasuryPool(_treasuryPool);
+  }
+
+  function updateWhitelistedTokens(address[] calldata _tokens, bool _isAdd)
+    external override onlyAdmin
+  {
+    _updateWhitelistedToken(_tokens, _isAdd);
+  }
+
+  /** @dev Liquidate list of tokens to a single dest token,
+  *   source token must not be a whitelisted token, dest must be a whitelisted token
+  * @param sources list of source tokens to liquidate
+  * @param amounts list of amounts corresponding to each source token
+  * @param recipient receiver of sources token
+  * @param dest token to liquidate to, must be whitelisted
+  * @param minReturn minimum return of dest token for this liquidation
+  * @param txData data to callback to recipient
+  */
+  function liquidate(
+    IERC20Ext[] memory sources,
+    uint256[] memory amounts,
+    address payable recipient,
+    IERC20Ext dest,
+    uint256 minReturn,
+    bytes memory txData
+  )
+    public override virtual onlyWhenLiquidationEnabled nonReentrant
+    returns (uint256 destAmount)
+  {
+    require(
+      whitelistedTokens[address(dest)],
+      'only liquidate to whitelisted tokens'
+    );
+    for(uint256 i = 0; i < sources.length; i++) {
+      require(
+        !whitelistedTokens[address(sources[i])],
+        'can not liquidate from a whitelisted token'
+      );
+    }
+    // withdraw funds from fee pool to recipient
+    feePool.withdrawFunds(sources, amounts, recipient);
+    uint256 balanceDestBefore = dest.balanceOf(treasuryPool);
```

Comment:
Feel safer if `feePool.withdrawFunds` is called after `dest.balanceOf(treasuryPool)`